Simple module to install Perl libraries via cpanm.

fcuny opened the pull request:
cpanm is a popular alternative to the CPAN client for installing Perl
libraries.  Unfortunately it can't uninstall already installed module, so it
can't handle states.

Commit:
Simple module to install Perl libraries via cpanm.

cpanm is a popular alternative to the CPAN client for installing Perl
libraries.  Unfortunately it can't uninstall already installed module, so it
can't handle states.

mpdehaan commented:
Thanks, made good use of cpanm (and Dancer, as in your examples) not long
ago.

(FYI -- This is in queue for the 1.0 release, since we are releasing 0.9
Friday)

Changes to make:

Name should probably take "pkg" as an alias to be consistent with other
modules like this.

It also looks like this should raise an error (fail_json) if from_path or
name isn't specified (one or the other should be required).

There is a facility for this in module common called required_one_of, which
takes a list, that can do this for you automatically.

Don't have of it in front of me to test, but I imagine it could support
removals, in which case it could take the state parameter?   Otherwise it
seems to be installing every time (or no)?

One thing we do want out of modules is some degree of idempotence against
making changes on repeated runs, if we can get them.

Is there anything that needs to be done to use this with regard to mirror
setup if a machine hasn't been configured to use CPAN yet?  I remember that
being another sore point, and perhaps is something to include in the module
docs.

About all I can think of at the moment, let me know your thoughts on the
above...

On Tue, Nov 27, 2012 at 7:33 PM, Franck Cuny notifications@github.comwrote:

> cpanm is a popular alternative to the CPAN client for installing Perl
> libraries. Unfortunately it can't uninstall already installed module, so it
> 
> ## can't handle states.
> 
> You can merge this Pull Request by running:
> 
>   git pull https://github.com/franckcuny/ansible module-cpanm
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/ansible/ansible/pull/1700
> Commit Summary
> - Simple module to install Perl libraries via cpanm.
> 
> File Changes
> - _A_ library/cpanm (118)
> 
> Patch Links
> - https://github.com/ansible/ansible/pull/1700.patch
> - https://github.com/ansible/ansible/pull/1700.diff
>   
>   —
>   Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/1700.

sfromm commented:
I'll also add that the parameter _notest_ should be made to use BOOLEANS.  See the seboolean module for an example.

Commit:
Add the `mirror` option and verify if a package is already installed.

By default, the cpanm client will use a default mirror to download
libraries, but it's possible for the end user to pass a different URL.

Since ansible favorite idempotence when possible, we verify if the
module is already installed before running the cpanm client.

Another minor additional change, the `notest` option is now a boolean.

fcuny commented:
> (FYI -- This is in queue for the 1.0 release, since we are releasing 0.9 
> Friday) 

That's fine with me.

> Changes to make: 
> 
> Name should probably take "pkg" as an alias to be consistent with other 
> modules like this. 
> 
> It also looks like this should raise an error (fail_json) if from_path or 
> name isn't specified (one or the other should be required). 
> 
> There is a facility for this in module common called required_one_of, which 
> takes a list, that can do this for you automatically. 

This function is already used (see line 81).  If there's no from_path or name, the result is:

{
    "failed": true, 
    "msg": "one of the following is required: ['name', 'from_path']"
}

> Don't have of it in front of me to test, but I imagine it could support 
> removals, in which case it could take the state parameter? Otherwise it 
> seems to be installing every time (or no)? 

sadly, no.  What I can do is to check if the module App::pmuninstall (http://search.cpan.org/~xaicron/App-pmuninstall-0.16/lib/App/pmuninstall.pm) is installed on the machine, and in that case allow the "absent" state and uninstall a module.

> One thing we do want out of modules is some degree of idempotence against 
> making changes on repeated runs, if we can get them. 

I'll add a function "is_present" to check if the module is already installed (and with an optional version parameter) and not run the cpanm client in that case.

> Is there anything that needs to be done to use this with regard to mirror 
> setup if a machine hasn't been configured to use CPAN yet? I remember that 
> being another sore point, and perhaps is something to include in the module 
> docs. 

By default cpanm is using a default mirror so you don't have to configure anything.  But it's true that it also accepts an optional parameter to use a specific mirror (that can be useful in the case of a private mirror with extra packages in it).  I'll add support for this option.

> About all I can think of at the moment, let me know your thoughts on the 
> above… 

Thanks!

> On Tue, Nov 27, 2012 at 7:33 PM, Franck Cuny notifications@github.comwrote: 
> 
> > cpanm is a popular alternative to the CPAN client for installing Perl 
> > libraries. Unfortunately it can't uninstall already installed module, so it 
> > 
> > ## can't handle states. 
> > 
> > You can merge this Pull Request by running: 
> > 
> > git pull https://github.com/franckcuny/ansible module-cpanm 
> > 
> > Or view, comment on, or merge it at: 
> > 
> > https://github.com/ansible/ansible/pull/1700 
> > Commit Summary 
> > - Simple module to install Perl libraries via cpanm. 
> > 
> > File Changes 
> > - _A_ library/cpanm (118) 
> > 
> > Patch Links 
> > - https://github.com/ansible/ansible/pull/1700.patch 
> > - https://github.com/ansible/ansible/pull/1700.diff 
> > 
> > — 
> > Reply to this email directly or view it on GitHubhttps://github.com/ansible/ansible/pull/1700. 
> > 
> > —
> > Reply to this email directly or view it on GitHub.

## 

Franck Cuny
franck@lumberjaph.net // http://lumberjaph.net
{twitter,github}/franckcuny

Commit:
Alias name to pkg.

mpdehaan commented:
I've decided to not to further expand core for bindings for every language, what I would suggest is you maintain this in it's own repo, and we can link to it from the resources page of the docs section

mpdehaan commented:
That decision was a while back and resources and such have changed a lot, so open to reconsidering this.   We may opt at tweaking this some as some things about imports have changed.

fcuny commented:
Hm OK, I'll take another look then. Thanks!

jctanner commented:
@franckcuny thanks!
